File: ts/project/.sdk/tm/dart/lib/feature/test/TestFeature.dart
// ignore_for_file: non_constant_identifier_names

import 'dart:async';
import 'dart:math';

import '../../utility/voxgig_struct.dart' as vs;

import '../base/BaseFeature.dart';

const S_NOT_FOUND = 'Not found';

class TestFeature extends BaseFeature {
  dynamic _client;
  int _netcalls = 0;

  TestFeature() {
    version = '0.0.1';
    name = 'test';
    active = true;
  }

  @override
  dynamic init(dynamic ctx, dynamic opts) {
    _client = ctx.client;
    options = opts is Map ? Map<String, dynamic>.from(opts) : {};

    final entity = options['entity'];

    _client.mode = 'test';

    // Ensure entity ids are correct.
    vs.walk(entity, before: (k, v, parent, path) {
      if (2 == vs.size(path)) {
        vs.setprop(v, 'id', k);
      }
      return v;
    });

    final self = this;

    dynamic respond(int status, [dynamic data, dynamic res]) {
      final out = vs.merge([
        <String, dynamic>{
          'status': status,
          'statusText': 'OK',
          'json': () => data,
          'body': 'not-used',
        },
        vs.getdef(res, {}),
      ]);

      final headers = vs.getprop(out, 'headers', {});
      out['headers'] = headers;

      return out;
    }

    dynamic testFetcher(dynamic fctx, dynamic _fullurl, dynamic _fetchdef) {
      final param = fctx.utility.param;

      final op = fctx.op;
      final entmap = vs.getprop(entity, op.entity, {});

      if ('load' == op.name) {
        final args = self.buildArgs(fctx, op, fctx.reqmatch);
        final found = vs.select(entmap, args);
        final ent = vs.getelem(found, 0);
        if (null == ent) {
          return respond(404, null, {'statusText': S_NOT_FOUND});
        } else {
          vs.delprop(ent, r'$KEY');
          final out = vs.clone(ent);
          return respond(200, out);
        }
      } else if ('list' == op.name) {
        final args = self.buildArgs(fctx, op, fctx.reqmatch);
        final found = vs.select(entmap, args);
        if (null == found) {
          return respond(404, null, {'statusText': S_NOT_FOUND});
        } else {
          for (final ent in found) {
            vs.delprop(ent, r'$KEY');
          }
          final out = vs.clone(found);
          return respond(200, out);
        }
      } else if ('update' == op.name) {
        final args = self.buildArgs(fctx, op, fctx.reqdata);
        final found = vs.select(entmap, args);
        final ent = vs.getelem(found, 0);
        if (null == ent) {
          return respond(404, null, {'statusText': S_NOT_FOUND});
        } else {
          // Dart's single null stands in for the donor's undefined: merge
          // must not overwrite stored values with absent ones.
          final upddata = <String, dynamic>{};
          if (fctx.reqdata is Map) {
            (fctx.reqdata as Map).forEach((k, v) {
              if (null != v) {
                upddata[k.toString()] = v;
              }
            });
          }
          vs.merge([ent, upddata]);
          vs.delprop(ent, r'$KEY');
          final out = vs.clone(ent);
          return respond(200, out);
        }
      } else if ('remove' == op.name) {
        final args = self.buildArgs(fctx, op, fctx.reqmatch);
        final found = vs.select(entmap, args);
        final ent = vs.getelem(found, 0);
        // Remove only the first matched entity. If nothing matches,
        // succeed as a no-op rather than erroring.
        if (null != ent) {
          vs.delprop(entmap, vs.getprop(ent, 'id'));
        }
        return respond(200);
      } else if ('create' == op.name) {
        self.buildArgs(fctx, op, fctx.reqdata);
        dynamic id = param(fctx, 'id');
        if (null == id) {
          final rng = Random();
          final h = () => rng.nextInt(0x10000).toRadixString(16);
          id = (h() + h() + h() + h()).padRight(16, '0');
        }

        final ent = vs.clone(fctx.reqdata);
        vs.setprop(ent, 'id', id);
        vs.setprop(entmap, id, ent);
        vs.delprop(ent, r'$KEY');
        final out = vs.clone(ent);
        return respond(200, out);
      }

      return null;
    }

    // Optional network behaviour simulation over the mock transport. Enable
    // per test via `SDK.test({ net: { latency, failTimes, ... } })`. When
    // `net` is absent the mock behaves exactly as before (no wrapping), so
    // existing generated tests are unaffected.
    final net = options['net'];
    ctx.utility.fetcher =
        (null == net) ? testFetcher : makeNetsim(net, testFetcher);
    return null;
  }

  // Wrap a transport with simulated network conditions: latency (fixed or
  // {min,max}), a budget of first-N failures (`failTimes` -> `failStatus`),
  // first-N connection errors (`errorTimes`), or a hard `offline` outage.
  // Counter-driven, so simulations are deterministic across a test.
  dynamic makeNetsim(dynamic net, dynamic inner) {
    final self = this;
    self._netcalls = 0;

    int pickLatency() {
      final l = net['latency'];
      if (null == l) {
        return 0;
      }
      if (l is num) {
        return l < 0 ? 0 : l.toInt();
      }
      final min = (l['min'] is num) ? (l['min'] as num).toInt() : 0;
      final max = null == l['max'] ? min : (l['max'] as num).toInt();
      return max <= min ? min : min + ((max - min) >> 1);
    }

    Future<void> sleep(int ms) {
      if (0 >= ms) {
        return Future.value();
      }
      final sleeper = net['sleep'];
      if (sleeper is Function) {
        return Future.value(sleeper(ms)).then((_x) {});
      }
      return Future.delayed(Duration(milliseconds: ms));
    }

    return (dynamic fctx, dynamic url, dynamic fetchdef) async {
      self._netcalls++;
      final call = self._netcalls;

      if (true == net['offline']) {
        await sleep(pickLatency());
        return fctx.error('netsim_offline',
            'Simulated network offline (URL was: "' + url.toString() + '")');
      }
      final errorTimes = net['errorTimes'] is num ? net['errorTimes'] : 0;
      if (call <= errorTimes) {
        await sleep(pickLatency());
        return fctx.error('netsim_conn',
            'Simulated connection error (call ' + call.toString() + ')');
      }
      final failTimes = net['failTimes'] is num ? net['failTimes'] : 0;
      if (call <= failTimes) {
        await sleep(pickLatency());
        final status = null == net['failStatus'] ? 503 : net['failStatus'];
        return <String, dynamic>{
          'status': status,
          'statusText': 'Simulated Failure',
          'body': 'not-used',
          'json': () => null,
          'headers': {},
        };
      }
      await sleep(pickLatency());
      return inner(fctx, url, fetchdef);
    };
  }

  dynamic buildArgs(dynamic ctx, dynamic op, dynamic args) {
    final param = ctx.utility.param;

    final opname = op.name;
    final point = vs.getelem(
        vs.getpath(ctx.config,
            ['entity', _entname(ctx.entity), 'op', opname, 'points']),
        -1);

    final reqd = vs.transform(
      vs.select(vs.getpath(point, ['args', 'params']), {'reqd': true}),
      ['`\$EACH`', '', '`\$KEY.name`'],
    );

    final qand = <dynamic>[];
    final q = {'`\$AND`': qand};

    for (final k in vs.keysof(args)) {
      if ('id' == k || !vs.isempty(vs.select(reqd, k))) {
        final v = param(ctx, k);
        // Dart's single null stands in for the donor's undefined: an
        // absent match value constrains nothing.
        if (null == v) {
          continue;
        }
        // NOTE: mirrors the donor: op.alias is not present on Operation,
        // so no alias alternates are generated here.
        dynamic qor = [
          {k: v}
        ];

        qor = {'`\$OR`': qor};

        qand.add(qor);
      }
    }

    if (null != ctx.ctrl['explain']) {
      ctx.ctrl['explain']['test'] = {'query': q};
    }

    return q;
  }

  String _entname(dynamic ent) {
    if (null == ent) {
      return '';
    }
    if (ent is Map) {
      return (vs.getprop(ent, 'name', '') ?? '').toString();
    }
    try {
      return (ent.name ?? '').toString();
    } catch (_e) {
      return '';
    }
  }
}

File: ts/project/.sdk/tm/ts/src/feature/test/TestFeature.ts

import type { Context, FeatureOptions } from '../../types'
import type { ProjectNameSDK } from '../../ProjectNameSDK'

import { BaseFeature } from '../base/BaseFeature'


const S_NOT_FOUND = 'Not found'


function ownIdField(config: any, getpath: any, entityName: string): string {
  let fallback = ''

  for (const opname of ['load', 'remove', 'update']) {
    const points = getpath(config, ['entity', entityName, 'op', opname, 'points']) || []
    const canonical = points.filter((pt: any) =>
      null == (pt && pt.select && pt.select['$action']))
    const use = 0 < canonical.length ? canonical : points
    let best = use[0]
    for (const pt of use) {
      if (null == pt || null == pt.parts || null == best || null == best.parts) continue
      const ptterm = 0 < pt.parts.length && String(pt.parts[pt.parts.length - 1]).startsWith('{')
      const bestterm = 0 < best.parts.length && String(best.parts[best.parts.length - 1]).startsWith('{')
      if (ptterm !== bestterm ? ptterm : pt.parts.length < best.parts.length) best = pt
    }
    const parts: string[] = (best && best.parts) || []
    const lastPart = 0 < parts.length ? String(parts[parts.length - 1]) : ''
    if (lastPart.startsWith('{')) return lastPart.slice(1, -1)

    // No path param at all (or the route ends in a literal, e.g.
    // /orgs/{org}/private-registries/public-key): a single required QUERY
    // param can still be the record's own key (e.g. GET /result?trace_id=).
    const query = (best && best.args && best.args.query) || []
    const reqdQuery = query.filter((q: any) => false !== q.reqd)
    if (1 === reqdQuery.length) return String(reqdQuery[0].name)

    // The last path parameter of a literal-terminal route is remembered as
    // a LAST RESORT, but not returned yet — see below.
    if ('' === fallback) {
      for (let i = parts.length - 1; 0 <= i; i--) {
        const part = String(parts[i])
        if (part.startsWith('{')) {
          fallback = part.slice(1, -1)
          break
        }
      }
    }
  }

  if ('' !== fallback) {
    return fallback
  }

  return 'id'
}


class TestFeature extends BaseFeature {
  version = '0.0.1'
  name = 'test'
  active = true

  _client?: ProjectNameSDK
  _options?: any


  init(ctx: Context, options: FeatureOptions): void | Promise<any> {
    this._client = ctx.client
    this._options = options

    const struct = ctx.utility.struct
    const walk = struct.walk
    const size = struct.size
    const setprop = struct.setprop

    const entity = this._options.entity

    this._client._mode = 'test'

    const getpath = struct.getpath

    // Ensure entity ids are correct.
    walk(entity, (k: any, v: any, _parent: any, path: any) => {
      if (2 === size(path)) {
        setprop(v, 'id', k)
        const idField = ownIdField(ctx.config, getpath, String(path[0]))
        if ('id' !== idField) {
          setprop(v, idField, k)
        }
      }
      return v
    })

    const self = this

    function testFetcher(ctx: any, _fullurl: string, _fetchdef: any) {
      const struct = ctx.utility.struct
      const param = ctx.utility.param

      const getprop = struct.getprop
      const clone = struct.clone
      const merge = struct.merge
      const getelem = struct.getelem
      const select = struct.select
      const delprop = struct.delprop
      const getdef = struct.getdef

      function envelope(data: any) {
        const restf = getprop(getprop(ctx.point, 'transform', {}), 'res')
        if (null == data || 'string' !== typeof restf) {
          return data
        }
        const m = restf.match(/^`body\.(.+)`$/)
        if (null == m) {
          return data
        }
        let out: any = data
        const segs = m[1].split('.')
        for (let i = segs.length - 1; 0 <= i; i--) {
          out = { [segs[i]]: out }
        }
        return out
      }

      function respond(status: number, data?: any, res?: any) {
        const payload = envelope(data)
        const out = merge([
          {
            status,
            statusText: 'OK',
            json: async () => payload,
            body: 'not-used',
          },
          getdef(res, {})
        ])

        const headers: any = getprop(out, 'headers', {})

        // JS specific iterator.
        out.headers = {
          forEach(callback: any) {
            Object.keys(headers).forEach((key) => {
              callback(headers[key], key, this)
            })
          }
        }

        return out
      }


      const op = ctx.op
      const entmap = getprop(entity, op.entity, {})

      if ('load' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        if (null == ent) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          delprop(ent, '$KEY')
          const out = clone(ent)
          return respond(200, out)
        }
      }
      else if ('list' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        if (null == found) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          found.map((ent: any) => delprop(ent, '$KEY'))
          const out = clone(found)
          return respond(200, out)
        }
      }
      else if ('update' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqdata)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        if (null == ent) {
          return respond(404, undefined, { statusText: S_NOT_FOUND })
        }
        else {
          merge([ent, (ctx.reqdata || {})])
          delprop(ent, '$KEY')
          const out = clone(ent)
          return respond(200, out)
        }
      }
      else if ('remove' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqmatch)
        const found = select(entmap, args)
        const ent = getelem(found, 0)
        // Remove only the first matched entity. If nothing matches,
        // succeed as a no-op rather than erroring.
        if (null != ent) {
          delprop(entmap, getprop(ent, 'id'))
        }
        return respond(200)
      }
      else if ('create' === op.name) {
        const args = self.buildArgs(ctx, op, ctx.reqdata)
        let id = param(ctx, 'id')
        if (null == id) {
          id = ((1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16) +
            (1e4 * Math.random() | 0).toString(16)).padEnd(16, '0')
        }

        const ent = clone(ctx.reqdata)
        setprop(ent, 'id', id)

        // A record created during the run needs the same real-key seeding
        // the initial walk gives seed data (see ownIdField above) — without
        // it, only `id` is set, and a load by the entity's own key right
        // after create (recordKey !== 'id') finds nothing.
        const idField = ownIdField(ctx.config, struct.getpath, getprop(op, 'entity'))
        if ('id' !== idField && null == getprop(ent, idField)) {
          setprop(ent, idField, id)
        }

        setprop(entmap, id, ent)
        delprop(ent, '$KEY')
        const out = clone(ent)
        return respond(200, out)
      }
    }

    // Optional network behaviour simulation over the mock transport. Enable
    // per test via `SDK.test({ net: { latency, failTimes, ... } })`. When
    // `net` is absent the mock behaves exactly as before (no wrapping), so
    // existing generated tests are unaffected.
    const net = this._options.net
    ctx.utility.fetcher = (null == net) ? testFetcher : this.makeNetsim(net, testFetcher)
  }


  // Wrap a transport with simulated network conditions: latency (fixed or
  // {min,max}), a budget of first-N failures (`failTimes` -> `failStatus`),
  // first-N connection errors (`errorTimes`), or a hard `offline` outage.
  // Counter-driven, so simulations are deterministic across a test.
  makeNetsim(this: any, net: any, inner: any) {
    const self = this
    self._netcalls = 0

    function pickLatency(): number {
      const l = net.latency
      if (null == l) { return 0 }
      if ('number' === typeof l) { return l < 0 ? 0 : l }
      const min = l.min | 0
      const max = null == l.max ? min : l.max | 0
      return max <= min ? min : min + ((max - min) >> 1)
    }

    function sleep(ms: number): Promise<void> {
      if (null == ms || 0 >= ms) { return Promise.resolve() }
      if ('function' === typeof net.sleep) { return Promise.resolve(net.sleep(ms)) }
      return new Promise((r) => setTimeout(r, ms))
    }

    return async function netsimFetcher(ctx: any, url: string, fetchdef: any) {
      self._netcalls++
      const call = self._netcalls

      if (true === net.offline) {
        await sleep(pickLatency())
        return ctx.error('netsim_offline', 'Simulated network offline (URL was: "' + url + '")')
      }
      if (call <= (net.errorTimes | 0)) {
        await sleep(pickLatency())
        return ctx.error('netsim_conn', 'Simulated connection error (call ' + call + ')')
      }
      if (call <= (net.failTimes | 0)) {
        await sleep(pickLatency())
        const status = null == net.failStatus ? 503 : net.failStatus
        return {
          status,
          statusText: 'Simulated Failure',
          body: 'not-used',
          json: async () => undefined,
          headers: { forEach(_cb: any) { }, get(_k: string) { return undefined } },
        }
      }
      await sleep(pickLatency())
      return inner(ctx, url, fetchdef)
    }
  }


  buildArgs(ctx: any, op: any, args: any): any {
    const struct = ctx.utility.struct
    const param = ctx.utility.param

    const getprop = struct.getprop
    const keysof = struct.keysof
    const getpath = struct.getpath
    const getelem = struct.getelem
    const select = struct.select
    const transform = struct.transform
    const isempty = struct.isempty

    const opname = getprop(op, 'name')
    const points = getpath(ctx.config, [
      'entity', getprop(ctx.entity, 'name'), 'op', opname, 'points']) || []

    // Pick the entity's own endpoint, not a cross-reference from another
    // resource that also returns it — the same rule makePoint falls back to,
    // so the seed-data query is built from the endpoint the request will
    // actually be sent to: a terminal `{id}` marks a record route, and
    // failing that the shallower path wins.
    const isterm = (pt: any) => {
      const parts = pt.parts
      const last = 0 < parts.length ? parts[parts.length - 1] : ''
      return 'string' === typeof last && 0 === last.indexOf('{')
    }

    let point = getelem(points, 0)
    for (let i = 1; i < points.length; i++) {
      const cand = getelem(points, i)
      if (isterm(cand) !== isterm(point) ? isterm(cand) :
        cand.parts.length < point.parts.length) {
        point = cand
      }
    }

    // Path AND query: a path-only read misses a query-addressed record
    // (e.g. GET /result?trace_id=), which has no path param at all.
    const reqdParams = transform(
      select(getpath(point, ['args', 'params']), { reqd: true }),
      ['`$EACH`', '', '`$KEY.name`']
    )
    const reqdQuery = transform(
      select(getpath(point, ['args', 'query']), { reqd: true }),
      ['`$EACH`', '', '`$KEY.name`']
    )
    const reqd = [...(reqdParams || []), ...(reqdQuery || [])]

    const qand: any[] = []
    const q = { '`$AND`': qand }

    const idfrom = getpath(ctx.config, [
      'entity', getprop(ctx.entity, 'name'), 'id', 'from']) || {}

    const nest = (path: string, value: any) => {
      const keys = String(path).split('.')
      let out: any = value
      for (let i = keys.length - 1; 0 <= i; i--) {
        out = { [keys[i]]: out }
      }
      return out
    }

    const match = ctx.match || {}
    const names: any[] = [...keysof(args)]

    if (null == getprop(args, 'id')) {
      for (const part of (getpath(ctx.config, [
        'entity', getprop(ctx.entity, 'name'), 'id', 'parts']) || [])) {
        const pname = String(part)
        if (!names.includes(pname) && null != getprop(match, pname)) {
          names.push(pname)
        }
      }
    }

    for (let k of names) {
      if ('id' === k || !isempty(select(reqd, k))) {
        const v = param(ctx, k)
        const ka = getprop(op.alias, k)

        let qor: any = [{ [k]: v }]
        if (null != ka) {
          qor.push({ [ka]: v })
        }

        // An ALTERNATIVE, never a replacement: the parameter name still
        // matches a record that happens to carry it flat, so a seed written
        // either way is found. `$OR` is what makes that safe.
        const kf = getprop(idfrom, k)
        if (null != kf && k !== kf) {
          qor.push(nest(kf, v))
        }

        qor = { '`$OR`': qor }

        qand.push(qor)
      }
    }

    if (ctx.ctrl.explain) {
      ctx.ctrl.explain.test = { query: q }
    }

    return q
  }
}


export {
  TestFeature,
  ownIdField,
}



